### PR TITLE
Disallow search request with preference parameter when weighted routing enabled

### PR DESCRIPTION
**Pull request by anshu1106**
Signed-off-by: Anshu Agarwal <anshukag@amazon.com>

### Description
Disallow search requests with preference parameter in case of strict weighted routing

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/5873

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java [hunk 1]
```diff
@@ -796,4 +798,66 @@ public void testMultiGetWithNetworkDisruption_FailOpenDisabled() throws Exceptio
...
+        int numReplicas = 1;
+        setUpIndexing(numShards, numReplicas);
+
+        logger.info("--> setting shard routing weights for weighted round robin");
+        Map<String, Double> weights = Map.of("a", 1.0, "b", 1.0, "c", 0.0);
+        setShardRoutingWeights(weights);
+
+        assertThrows(
+            PreferenceBasedSearchNotAllowedException.class,
+            () -> internalCluster().client(nodeMap.get("b").get(0)).prepareSearch().setSize(0).setPreference("_only_local").get()
```
Comment (Bukhtawar) [inline server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java:827]: can we add randomBetween(local, node_id, only_node) jus to cover all preference options

Comment (anshu1106) [reply server/src/internalClusterTest/java/org/opensearch/search/SearchWeightedRoutingIT.java]: addressed

**Review by Bukhtawar — APPROVED** [review]

**Comment by kavilla** [comment]
@Bukhtawar, @dblock, @nknize.

This is technically a breaking change correct? From OpenSearch Dashboards perspective, by default we pass the value from OpenSearch on the request with a session ID say that restricts operations to execute all search requests on the same shards. This has the benefit of reusing shard caches across requests. This helps ensure the results are the same per a session based on their sharding strategy. By disallowing this and with our default setting not knowing when weighted routing is enabled then requests will fail.

But being that we are only one consumer of the `_search` API, it is an important one so I'd imagine folks who do not use OpenSearch Dashboards but utilize OpenSearch and would like to enable weighted routing then they have to make sure they will have to remove the `preference` param from their request.

So I understand if this is in `main` but having it be backported to `2.x` and `1.x` seems to break SEMVER to me if the weighted routing where released on those major versions. If there is an improvement to be made about weighted routing could it respect both the weights and the preference because it seems like to ignore a requests preference seems not as user friendly.

Disregard if the weighted routing was not release in a previous version.

cc: @joshuarrrr

**Comment by andrross** [comment]
@kavilla Weighted routing was [introduced in 2.4](https://github.com/opensearch-project/OpenSearch/blob/main/release-notes/opensearch.release-notes-2.4.0.md?plain=1#L7), though it appears it wasn't documented [until 2.5](https://opensearch.org/docs/2.5/api-reference/cluster-api/cluster-awareness/). It does sound like this is a breaking change for 2.x. @Bukhtawar @anshu1106 What do you think?

**Comment by anshu1106** [comment]
Makes sense. Created an issue https://github.com/opensearch-project/OpenSearch/issues/6297 to address this.

**Pull request merged**